lib/recursive-methods.rb

```diff
@@ -1,49 +1,122 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n) b/c it's being called recursively n times before reaching the base case. (linear)
+# Space complexity: 0(n) There are no new variables but there is the Stack! 
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+	# 3! = 3 * 2 * 1 = 6
+	# 1. If...a recursive case (method calls itself)
+	# and a base case (method ends - does not call itself)
+	# 2 Handle simplest case first - base case( no looping)
+	# 3. Write recursive call (one step simpler each time (n-1)
+	# 4. Assume the call works. 
+	# ask myself...what does it do? 
+	# ask myself...how does it help? 
+	if !(n >= 0)
+		raise ArgumentError.new("The input must be greater than or equal to 0")
+	end
+	# base case
+		if n <= 1
+			return 1
+		else
+			# recursive case
+			return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: 0(n^2) It will always go the length of the string plus a new variable. 
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+	# base case
+		return s if s.length <= 1
+		# recursive case - with range argument found here:  https://ruby-doc.org/core-2.5.0/Range.html
+		reversed_str = reverse(s[1..-1])
+		# "t", "ta", "tac"
+		# new string that contains the first reversed character
+		reversed_str << s[0]
+		reversed_str
+	
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - n being the length
+# Space complexity: 0(n) - No temporary variables so 0(1),  but there is the Stack so 0(n)??
 def reverse_inplace(s)
```

Comment:
Since you are making a new string with each recursive call this is O(n^2) space complexity.

This also means you're not doing it in-place.  You're actually making a new array with each recursive call.